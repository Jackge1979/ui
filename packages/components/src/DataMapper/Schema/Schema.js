import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SchemaElement from './SchemaElement/SchemaElement.js';
import DraggableSchemaElement from './SchemaElement/DraggableSchemaElement.js';

function isMapped(element, mapped) {
	return mapped == null ? false : mapped.includes(element);
}

function isSelected(element, selection, type) {
	return selection == null ?
		false	: (selection.type === type && selection.element === element);
}

function isHighlighted(element, selection, type, pendingItem, focusedElements) {
	const connected = selection == null ?
	 false : (selection.type !== type
		 && selection.connected != null
		 && selection.connected.includes(element));
	const pending = pendingItem != null
			&& pendingItem.type === type
			&& pendingItem.element === element;
	const focused = focusedElements != null && focusedElements.includes(element);
	return connected || pending || focused;
}

function renderSchemaElement(type, elem, draggable, mapped, performMapping,
				selection, onSelect, pendingItem, onEnterElement, onLeaveElement,
				focusedElements) {
	if (draggable) {
		return (
			<DraggableSchemaElement
				key={elem}
				name={elem}
				schemaType={type}
				mapped={isMapped(elem, mapped)}
				performMapping={performMapping}
				selected={isSelected(elem, selection, type)}
				highlighted={isHighlighted(elem, selection, type, pendingItem, focusedElements)}
				onSelect={onSelect}
				onEnterElement={onEnterElement}
				onLeaveElement={onLeaveElement}
			/>
		);
	}
	return (
		<SchemaElement
			key={elem}
			name={elem}
			schemaType={type}
			selected={isSelected(elem, selection, type)}
			highlighted={isHighlighted(elem, selection, type, pendingItem, focusedElements)}
			onSelect={onSelect}
			onEnterElement={onEnterElement}
			onLeaveElement={onLeaveElement}
		/>
	);
}

export default class Schema extends Component {

	getNode(element) {
		const children = this.contentNode.childNodes;
		const childranArray = Array.from(children);
		return childranArray.find(c => c.firstChild.innerHTML === element);
	}

	getYPosition(element) {
		const scrollTop = this.contentNode.scrollTop;
		const child = this.getNode(element);
		const childOffsetTop = child.offsetTop;
		const parentOffsetTop = this.contentNode.offsetTop;
		const y = childOffsetTop - parentOffsetTop + child.clientHeight / 2 - scrollTop;
		return y;
	}

	reveal(element) {
		const node = this.getNode(element);
		const nodeHeight = node.clientHeight;
		const elemYPos = this.getYPosition(element);
		const contentHeight = this.contentNode.offsetHeight;
		if (elemYPos < 0) {
			this.contentNode.scrollTop = this.contentNode.scrollTop + elemYPos - nodeHeight / 2;
		} else if (elemYPos > contentHeight - nodeHeight) {
			const offset = elemYPos + nodeHeight - contentHeight;
			this.contentNode.scrollTop = this.contentNode.scrollTop + offset;
		}
	}

	render() {
		const { type, schema, draggable, mapped, performMapping,
			selection, onSelect, pendingItem,
			onEnterElement, onLeaveElement, focusedElements } = this.props;
		return (
			<div className="schema mapper-element">
				<div className="schema-name">{schema.name}</div>
				<div
					ref={content => {
						this.contentNode = content;
					}}
					className="schema-content"
					onScroll={this.props.onScroll}
				>
					{schema.elements.map(elem =>
						renderSchemaElement(type, elem, draggable, mapped, performMapping,
							selection, onSelect, pendingItem, onEnterElement, onLeaveElement,
							focusedElements),
					)}
				</div>
			</div>
		);
	}
}

Schema.propTypes = {
	type: PropTypes.string,
	schema: PropTypes.object,
	mapped: PropTypes.array,
	performMapping: PropTypes.func,
	selection: PropTypes.object,
	draggable: PropTypes.bool,
	onSelect: PropTypes.func,
	onScroll: PropTypes.func,
	pendingItem: PropTypes.object,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	focusedElements: PropTypes.array,
};
