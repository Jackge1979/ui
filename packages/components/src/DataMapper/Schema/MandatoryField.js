import React from 'react';
import PropTypes from 'prop-types';

/**
 * This component displays a data as a label with mandatory info.
 * The data must be an object { value: string, mandatory: boolean }.
 */
export default function MandatoryField({ data, className }) {
	const label = data.value;
	const mandatory = data.mandatory;
	return (
		<div className={`tc-table-mandatory-field ${className}`}>
			<div className={'tc-table-mandatory-field-label'}>{label}</div>
			<div className={'tc-table-mandatory-field-info'}>{mandatory ? '*' : ''}</div>
		</div>
	);
}

MandatoryField.propTypes = {
	data: PropTypes.object,
	className: PropTypes.string,
};
