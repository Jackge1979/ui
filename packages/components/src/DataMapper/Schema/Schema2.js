import React, { Component } from 'react';
import PropTypes from 'prop-types';

function isMapped(dataAccessor, element, mappedElements) {
	return mappedElements == null ?
		false : dataAccessor.includes(mappedElements, element);
}

/**
 * isSelected indicates if the given (element, side) is selected
 * (i.e. if it appears in the selection)
 */
export function isSelected(dataAccessor, selection, element, side) {
	return selection != null
		&& dataAccessor.areEquals(selection.element, element)
		&& selection.side === side;
}

function isHighlighted(dataAccessor, element, selection, side, pendingItem, focusedElements) {
	const connected =
		selection == null
			? false
			: selection.side !== side &&
			  selection.connected != null &&
			  dataAccessor.includes(selection.connected, element);
	const pending =
		pendingItem != null
		&& pendingItem.side === side
		&& dataAccessor.areEquals(pendingItem.element, element);
	const focused = focusedElements != null
		&& dataAccessor.includes(focusedElements, element);
	return connected || pending || focused;
}

export default class Schema2 extends Component {

  constructor(props) {
		super(props);
		this.updateContentNodeRef = this.updateContentNodeRef.bind(this);
	}

	shouldComponentUpdate(nextProps) {
		return !(nextProps.dnd != null && (
			nextProps.dnd.pos != null || (
				nextProps.dnd.source != null && nextProps.dnd.source.side === nextProps.side
			)
		));
	}

	getNode(element) {
		const index = this.props.dataAccessor.getSchemaElementIndex(this.props.schema, element);
		const children = this.contentNode.childNodes;
		const childrenArray = Array.from(children);
		return childrenArray[index];
	}

	getYPosition(element) {
		const scrollTop = this.contentNode.scrollTop;
		const child = this.getNode(element);
		const childOffsetTop = child.offsetTop;
		const parentOffsetTop = this.contentNode.offsetTop;
		const y = childOffsetTop - parentOffsetTop + child.clientHeight / 2 - scrollTop;
		return y;
	}

	getVisibleElements() {
		let visibleElements = [];
		if (this.contentNode) {
			const contentHeight = this.contentNode.offsetHeight;
			const elements = this.props.dataAccessor.getSchemaElements(this.props.schema);
			const children = this.contentNode.childNodes;
			const childrenArray = Array.from(children);
			for (let i = 0; i < childrenArray.length; i += 1) {
				const element = elements[i];
				const elemYPos = this.getYPosition(element);
				if (elemYPos > 0 && elemYPos < contentHeight) {
					// element is visible
					visibleElements = visibleElements.concat(element);
				} else if (elemYPos > contentHeight) {
					break;
				}
			}
		}
		return visibleElements;
	}

	reveal(element) {
		const node = this.getNode(element);
		const nodeHeight = node.clientHeight;
		const elemYPos = this.getYPosition(element);
		const contentHeight = this.contentNode.offsetHeight;
		if (elemYPos < 0) {
			this.contentNode.scrollTop = this.contentNode.scrollTop + elemYPos - nodeHeight / 2;
		} else if (elemYPos > contentHeight - nodeHeight) {
			const offset = elemYPos + nodeHeight - contentHeight;
			this.contentNode.scrollTop = this.contentNode.scrollTop + offset;
		}
	}

	updateContentNodeRef(ref) {
		this.contentNode = ref;
	}

	render() {
    const {
      schemaRenderer,
      ...tempProps,
    } = this.props;
		const {
			dataAccessor,
			schema,
			side,
		} = this.props;
    const contentProps = {
      ...tempProps,
      updateContentNodeRef: this.updateContentNodeRef,
      isMapped,
      isSelected,
      isHighlighted,
    };
		return (
			<div className={`schema mapper-element ${side}`}>
				<div className="schema-name">{dataAccessor.getSchemaName(schema)}</div>
				{schemaRenderer.renderContent(contentProps)}
			</div>
		);
	}
}

Schema2.propTypes = {
	dataAccessor: PropTypes.object,
	schema: PropTypes.object,
	schemaRenderer: PropTypes.object,
};
