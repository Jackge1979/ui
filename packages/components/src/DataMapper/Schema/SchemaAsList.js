import React, { Component } from 'react';
import PropTypes from 'prop-types';
import VirtualizedList from '../../VirtualizedList';

export default class SchemaAsList extends Component {

	constructor(props) {
		super(props);
	}

	getYPosition(element) {
		return 0;
	}

	getVisibleElements() {
		let visibleElements = [];
		return visibleElements;
	}

	reveal(element) {
		console.log('reveal(' + element + ')');
	}

	updateContentNodeRef(ref) {
		this.contentNode = ref;
	}

	render() {
		const {
			type,
			schema,
			draggable,
			mapped,
			performMapping,
			selection,
			onSelect,
			pendingItem,
			onEnterElement,
			onLeaveElement,
			focusedElements,
			beginDrag,
			canDrop,
			drop,
			endDrag,
			revealConnection,
		} = this.props;
		return (
			<div className="schema mapper-element">
				<div className="schema-name">{schema.name}</div>
				<div
					ref={this.updateContentNodeRef}
					className="schema-content"
					onScroll={this.props.onScroll}
				>
					
				</div>
			</div>
		);
	}
}

Schema.propTypes = {
	type: PropTypes.string,
	schema: PropTypes.object,
	mapped: PropTypes.array,
	performMapping: PropTypes.func,
	selection: PropTypes.object,
	draggable: PropTypes.bool,
	onSelect: PropTypes.func,
	onScroll: PropTypes.func,
	pendingItem: PropTypes.object,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	focusedElements: PropTypes.array,
	beginDrag: PropTypes.func,
	canDrop: PropTypes.func,
	drop: PropTypes.func,
	endDrag: PropTypes.func,
	revealConnection: PropTypes.func,
};
