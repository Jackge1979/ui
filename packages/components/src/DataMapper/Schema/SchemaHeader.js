import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SchemaName from './SchemaName';

function renderFilter(filter, filterComponents, onFilterChange, side) {
	const FilterComponent = filterComponents.get(filter.getId());
	if (FilterComponent) {
		return (
			<FilterComponent
				key={`${filter.getId()}-${side}`}
				filter={filter}
				onFilterChange={onFilterChange}
			/>
		);
	}
	return <div>{`Filter ${filter.getId()} cannot be rendered`}</div>;
}

export default class SchemaHeader extends Component {
	constructor(props) {
		super(props);
		this.onFilterChange = this.onFilterChange.bind(this);
	}

	onFilterChange(filter) {
		this.props.onFilterChange(this.props.side, filter);
	}

	render() {
		const { dataAccessor, schema, side, filters, filterComponents } = this.props;
		return (
			<div className={`schema-header ${side}`}>
				<SchemaName dataAccessor={dataAccessor} schema={schema} side={side} />
				{
					filters.map(filter =>
						renderFilter(filter, filterComponents, this.onFilterChange, side))
				}
			</div>
		);
	}
}

SchemaHeader.propTypes = {
	dataAccessor: PropTypes.object,
	schema: PropTypes.object,
	side: PropTypes.string,
	filters: PropTypes.array,
	filterComponents: PropTypes.object,
	onFilterChange: PropTypes.func,
};
