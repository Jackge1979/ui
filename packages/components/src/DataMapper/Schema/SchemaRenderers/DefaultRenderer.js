import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SchemaElement from '../SchemaElement/SchemaElement.js';
import DraggableSchemaElement from '../SchemaElement/DraggableSchemaElement.js';

function renderSchemaElement(props, element) {
	const {
		dataAccessor,
		side,
		draggable,
		mappedElements,
		selection,
		onSelect,
		pendingItem,
		onEnterElement,
		onLeaveElement,
		focusedElements,
		dndListener,
		revealConnectedElement,
		isMapped,
		isSelected,
		isHighlighted,
	} = props;
	const key = dataAccessor.getElementId(element);
	const selected = isSelected(dataAccessor, selection, element, side);
	const highlighted = isHighlighted(
		dataAccessor,
		element,
		selection,
		side,
		pendingItem,
		focusedElements,
	);
	if (draggable) {
		return (
			<DraggableSchemaElement
				dataAccessor={dataAccessor}
				key={key}
				element={element}
				side={side}
				mapped={isMapped(dataAccessor, element, mappedElements)}
				selected={selected}
				highlighted={highlighted}
				onSelect={onSelect}
				onEnterElement={onEnterElement}
				onLeaveElement={onLeaveElement}
				dndListener={dndListener}
				revealConnectedElement={revealConnectedElement}
			/>
		);
	}
	return (
		<SchemaElement
			dataAccessor={dataAccessor}
			key={key}
			element={element}
			side={side}
			selected={selected}
			highlighted={highlighted}
			onSelect={onSelect}
			onEnterElement={onEnterElement}
			onLeaveElement={onLeaveElement}
			revealConnectedElement={revealConnectedElement}
		/>
	);
}

renderSchemaElement.propTypes = {
	dataAccessor: PropTypes.object,
	side: PropTypes.string,
	mappedElements: PropTypes.array,
	selection: PropTypes.object,
	draggable: PropTypes.bool,
	onSelect: PropTypes.func,
	pendingItem: PropTypes.object,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	focusedElements: PropTypes.array,
	dndListener: PropTypes.object,
	revealConnectedElement: PropTypes.func,
	isMapped: PropTypes.func,
	isSelected: PropTypes.func,
	isHighlighted: PropTypes.func,
};

export default class DefaultRenderer extends Component {
	render() {
		const { dataAccessor, schema, onScroll, updateContentNodeRef } = this.props;
		const content = dataAccessor
			.getSchemaElements(schema, true)
			.map(elem => renderSchemaElement(this.props, elem));
		return (
			<div ref={updateContentNodeRef} className="schema-content" onScroll={onScroll}>
				{content}
			</div>
		);
	}
}

DefaultRenderer.propTypes = {
	dataAccessor: PropTypes.object,
	schema: PropTypes.object,
	onScroll: PropTypes.func,
	updateContentNodeRef: PropTypes.func,
};
