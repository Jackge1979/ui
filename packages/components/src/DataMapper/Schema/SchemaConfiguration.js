import DefaultRenderer from './SchemaRenderers/DefaultRenderer';
import * as Constants from '../Constants';

export const inputColumns = [
  Constants.Schema.DATA_KEYS.TYPE,
	Constants.Schema.DATA_KEYS.NAME,
];

export const outputColumns = [
	Constants.Schema.DATA_KEYS.NAME,
  Constants.Schema.DATA_KEYS.TYPE,
	Constants.Schema.DATA_KEYS.DESC,
];

export default class SchemaConfiguration {

  getRenderer(side) {
    return DefaultRenderer;
  }

  getColumns(side) {
    return [];
  }

}
