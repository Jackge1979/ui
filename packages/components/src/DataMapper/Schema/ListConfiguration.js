import ListRenderer from './SchemaRenderers/ListRenderer';
import SchemaConfiguration, { inputColumns, outputColumns } from './SchemaConfiguration';
import * as Constants from '../Constants';

export default class ListConfiguration extends SchemaConfiguration {

  getRenderer(side) {
    return ListRenderer;
  }

  getColumns(side) {
    switch (side) {
      case Constants.MappingSide.INPUT:
        return inputColumns;
      case Constants.MappingSide.OUTPUT:
        return outputColumns;
      default:
        return [];
    }
  }

}
