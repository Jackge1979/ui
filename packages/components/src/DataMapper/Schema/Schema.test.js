import React from 'react';
import renderer from 'react-test-renderer';
import Schema from './Schema.js';
import DefaultDataAccessor from '../DefaultDataAccessor';
import DataAccessorWrapper from '../DataAccessorWrapper';
import DefaultRenderer from '../Schema/SchemaRenderers/DefaultRenderer';

const dataAccessor = new DataAccessorWrapper(new DefaultDataAccessor());

const emptySchema = {
	id: 'schema_1',
	name: 'schema',
	elements: [],
};

const element1 = {
	id: '1',
	name: 'element_1',
	type: 'string',
	description: 'bla bla bla',
};

const element2 = {
	id: '2',
	name: 'element_2',
	type: 'string',
	description: 'bla bla bla',
};

const element3 = {
	id: '3',
	name: 'element_3',
	type: 'string',
	description: 'bla bla bla',
};

const schema = {
	id: 'schema_2',
	name: 'schema',
	elements: [element1, element2, element3],
};

const noFilters = [];

it('single-schema', () => {
	// create React tree
	const tree = renderer
		.create(
			<Schema
				dataAccessor={dataAccessor}
				schema={schema}
				SchemaRenderer={DefaultRenderer}
				filters={noFilters}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('empty-schema', () => {
	// create React tree
	const tree = renderer
		.create(
			<Schema
				dataAccessor={dataAccessor}
				schema={emptySchema}
				SchemaRenderer={DefaultRenderer}
				filters={noFilters}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
