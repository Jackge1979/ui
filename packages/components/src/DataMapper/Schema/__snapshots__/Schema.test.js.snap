// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty-schema 1`] = `
<div
  className="schema mapper-element"
>
  <div
    className="schema-name"
  >
    schema
  </div>
  <div
    className="schema-content"
    onScroll={undefined}
  />
</div>
`;

exports[`single-schema 1`] = `
<div
  className="schema mapper-element"
>
  <div
    className="schema-name"
  >
    schema
  </div>
  <div
    className="schema-content"
    onScroll={undefined}
  >
    <div
      className="schema-element undefined"
      onClick={[Function]}
      onDoubleClick={[Function]}
      role="button"
    />
    <div
      className="schema-element undefined"
      onClick={[Function]}
      onDoubleClick={[Function]}
      role="button"
    />
    <div
      className="schema-element undefined"
      onClick={[Function]}
      onDoubleClick={[Function]}
      role="button"
    />
  </div>
</div>
`;
